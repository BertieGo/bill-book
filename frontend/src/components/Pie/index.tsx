import React from "react";
import {Tooltip} from "antd";
import isEqual from "lodash/isEqual";

import {IChartProps, IPathItem} from "../../declare";
import {fixedFloat, floorFloat} from "../../utils/math";


const FULL_CIRCLE_RADIUS = 360;

const HALF_CIRCLE_RADIUS = FULL_CIRCLE_RADIUS / 2;

export default class Pie extends React.Component<IChartProps, {}>{

    private width: number;

    private height: number;

    constructor(props: IChartProps) {
        super(props);
        this.width = 200;
        this.height = 200;
    }

    shouldComponentUpdate(nextProps: Readonly<IChartProps>, nextState: Readonly<{}>, nextContext: any): boolean {
        const { data } = this.props;
        return !isEqual(nextProps.data, data);
    }

    renderPath = () => {
        const { data } = this.props;
        const paths: JSX.Element[] = [];

        const radius = this.width / 2;

        const cx = this.width / 2;
        const cy = this.height / 2;

        let sx = this.width;
        let sy = this.height / 2;

        let prevPercent = 0;

        data.forEach((item: IPathItem) => {

            const currentPercent = (item.percent / 100) + prevPercent;

            const currentRadius = currentPercent * FULL_CIRCLE_RADIUS;

            const ex = floorFloat(cx + radius * Math.cos( currentRadius * Math.PI / HALF_CIRCLE_RADIUS));
            const ey = floorFloat(cy + radius * Math.sin( currentRadius * Math.PI / HALF_CIRCLE_RADIUS));

            const largeArcFlag = (item.percent / 100) * FULL_CIRCLE_RADIUS > HALF_CIRCLE_RADIUS ? '1' : '0';

            const displayPercent = `${floorFloat(item.percent)}%`;
            const displayAmount = fixedFloat(item.amount);

            paths.push(
                <Tooltip key={item._id} title={(<div className="nowrap">{`类型：${item.title} 金额: ${displayAmount} 占比：${displayPercent}`}</div>)}>
                    <path d={`M${sx} ${sy} A ${radius} ${radius}, 0, ${largeArcFlag}, 1, ${ex} ${ey} L ${cx} ${cy} Z`} fill={item.color}/>
                </Tooltip>
            )

            sx = ex;
            sy = ey;
            prevPercent = currentPercent;
        })

        return paths;
    }


    render() {
        return (
            <div>
                <svg width={`${this.width}px`} height={`${this.height}px`} version="1.1" xmlns="http://www.w3.org/2000/svg">
                    { this.renderPath() }
                </svg>
            </div>
        )
    }
}
