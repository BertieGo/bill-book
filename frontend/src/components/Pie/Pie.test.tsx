import React from 'react';
import Pie from './index';
import renderer from 'react-test-renderer';
import createReactAppJestPolyfill from '../../utils/test/polyfill';

createReactAppJestPolyfill();

test('COMPONENT-TEST: Pie', () => {
    const data = [{
        amount: 2672740,
        category: '8s0p77c323',
        color: "#72a935",
        percent: 55.4886,
        time: "2019-06-30T16:00:00.000Z",
        title: '支出',
        type: 0,
        _id: '5efdc622c42bdc4067a47c5c'
    }];

    const tree = renderer
        .create(<Pie data={data as any[]}/>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});


