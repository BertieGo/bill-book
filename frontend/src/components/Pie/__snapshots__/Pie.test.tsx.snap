// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT-TEST: CategorySelector 1`] = `
<div
  className="ant-select ant-select-single ant-select-allow-clear ant-select-show-arrow"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  style={
    Object {
      "width": 120,
    }
  }
>
  <div
    className="ant-select-selector"
    onClick={[Function]}
    onMouseDown={[Function]}
  >
    <span
      className="ant-select-selection-search"
    >
      <input
        aria-activedescendant="undefined_list_0"
        aria-autocomplete="list"
        aria-controls="undefined_list"
        aria-haspopup="listbox"
        aria-owns="undefined_list"
        autoComplete="off"
        className="ant-select-selection-search-input"
        onChange={[Function]}
        onCompositionEnd={[Function]}
        onCompositionStart={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onPaste={[Function]}
        readOnly={true}
        role="combobox"
        style={
          Object {
            "opacity": 0,
          }
        }
        unselectable="on"
        value=""
      />
    </span>
    <span
      className="ant-select-selection-placeholder"
    >
      请选择类型
    </span>
  </div>
  <span
    aria-hidden={true}
    className="ant-select-arrow"
    onMouseDown={[Function]}
    style={
      Object {
        "WebkitUserSelect": "none",
        "userSelect": "none",
      }
    }
    unselectable="on"
  >
    <span
      aria-label="down"
      className="anticon anticon-down"
      role="img"
    >
      <svg
        aria-hidden="true"
        className=""
        data-icon="down"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="64 64 896 896"
        width="1em"
      >
        <path
          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
        />
      </svg>
    </span>
  </span>
</div>
`;

exports[`COMPONENT-TEST: Pie 1`] = `
<div>
  <svg
    height="200px"
    version="1.1"
    width="200px"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M200 100 A 100 100, 0, 1, 1, 5.8876 66.1936 L 100 100 Z"
      fill="#72a935"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    />
  </svg>
</div>
`;
